File: Homework6/Program/Model/Parser/InfixParser.java
package Homework6.Program.Model.Parser;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class InfixParser {
    static Stack<String> rpn_out;  // output stack
    // operations priority
    static Map<String, Integer> opsPriority = new HashMap<>(Map.of(
            "(", 0, "-", 1, "+", 1, "/", 2, "*", 2, "^", 3));

    public static Stack<String> parse(Stack<String> tokens) {
        rpn_out = new Stack<>();
        Stack<String> operStack = new Stack<>(); // operations tokens stack
        while (!tokens.empty()) {
            String token = tokens.pop();
            switch (token) {
                // if open brace token, push it to operation stack
                case "(":
                    operStack.push(token);
                    break;
                // if close brace token, push all from operation stack to output stack until
                // open brace
                case ")":
                    while (operStack.size() > 0 && !operStack.peek().equals("(")) {
                        rpn_out.push(operStack.pop());
                    }
                    // and pop open brace
                    operStack.pop();
                    break;
                default:
                    // if operation token, push all higher or equal priority ops from operation stack
                    // to output stack
                    if (opsPriority.containsKey(token)) {
                        while (operStack.size() > 0
                                && opsPriority.get(operStack.peek()) >= opsPriority.get(token)) {
                            rpn_out.push(operStack.pop());
                        }
                        // push operation token to operation stack
                        operStack.push(token);
                        break;
                    }
                    // push number token to output stack
                    rpn_out.push(token);
                    break;
            }
        }
        // after all, push all remaining operations tokens to output stack
        while (!operStack.empty()) {
            rpn_out.push(operStack.pop());
        }
        Collections.reverse(rpn_out);
        return rpn_out;
    }

}

File: Homework7/model/InfixParser.py
from collections import deque
ops_priority = {"(": 0, "-": 1, "+": 1, "/": 2, "*": 2, "^": 3}
ops = ops = ['(', ')', '-', '+', '/', '*', '^']

rpn_expr = deque()


def parse(tokens: deque) -> deque:
    oper_stack = deque()
    while len(tokens) > 0:
        token = tokens.popleft()
        print(token)
        if token in ops:
            if token == ')':
                while len(oper_stack) > 0 and oper_stack[len(oper_stack)-1] != '(':
                    print(oper_stack[0])
                    print("logic", oper_stack[0] != '(')
                    rpn_expr.append(oper_stack.pop())
                    print("brace", oper_stack)
                    print("rpn", rpn_expr)
                print("over")
                oper_stack.pop()
            elif token == '(':
                oper_stack.append(token)
            else:
                while len(oper_stack) > 0 and \
                    ops_priority[oper_stack[len(oper_stack)-1]] >= ops_priority[token]:
                    rpn_expr.append(oper_stack.pop())
                oper_stack.append(token)
        else:
            rpn_expr.append(token)
        print("rpn", rpn_expr)
        print("opstack", oper_stack)
    rpn_expr.append(oper_stack.pop())
    return rpn_expr
